### PR TITLE
[FIX] website_sale_delivery: keep delivery method when using giftcards

### PR DESCRIPTION
**Pull request by robinengels**
Current behavior:
If you have atleast 2 delivery methods available on the website,
and try to pay with a giftcard the selected delivery method will
always switch to the first method of the list.

Steps to reproduce:
- Activate giftcards for website
- Publish atleast 2 delivery methods (one free and one not free)
- Go on the website shop, and add product to the cart
- Process checkout and select the not free delivery method
- Pay with giftcard
- The delivery price is 0 so the delivery method has been
  modified

opw-2883592
--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by nle-odoo — COMMENTED** [review]
the redirect /shop/payment was added in 2014 (bcc307560c35c9b9a7c0a01b0678fb9ac7dae3e2) but maybe it's not necessary anymore (and I guess it's necessary to not lose the gift_card_error)

as far as I can tell, you can have `website_sale_gift_card` installed without delivery (so the carrier_id field could not exist) so referring to carrier_id in it is not good

checking the `_check_carrier_quotation` method, if there is no set carrier, it seems we force the carrier only if we have set a particular carrier for a customer, so I'm not sure why the issue is happening, and if we were forcing a given carrier when showing the payment page, why it should not be forced anymore if the payment page is refreshed with a coupon

**Review by nle-odoo — APPROVED** [review]
looks to me (code wise)

in stable, I'm not sure we could add an argument to the method `_check_carrier_quotation` (the issue is that if someone overrode the method (which is possible in stable), his override would receive a method call with an unexpected argument) but we could transmit it through context

File: addons/website_sale_delivery/controllers/main.py [hunk 1]
```diff
@@ -13,10 +13,13 @@ class WebsiteSaleDelivery(WebsiteSale):
     def shop_payment(self, **post):
         order = request.website.sale_get_order()
         carrier_id = post.get('carrier_id')
+        keep_carrier = post.get('keep_carrier')
```
Comment (Williambraecky) [inline addons/website_sale_delivery/controllers/main.py:16]: I prefer if we default to False, avoiding None
```suggestion
        keep_carrier = post.get('keep_carrier', False)
```

File: addons/website_sale_delivery/models/sale_order.py [hunk 2]
```diff
@@ -28,7 +28,7 @@ def _compute_amount_delivery(self):
             else:
                 order.amount_delivery = sum(order.order_line.filtered('is_delivery').mapped('price_total'))
 
-    def _check_carrier_quotation(self, force_carrier_id=None):
+    def _check_carrier_quotation(self, force_carrier_id=None, ):
```
Comment (Williambraecky) [inline addons/website_sale_delivery/models/sale_order.py:31]: Unnecessary diff?
In master we can pass `keep_carrier` as argument to avoid overusing the context.

**Review by Williambraecky — CHANGES_REQUESTED** [review]
Asides from those two small comments the fix LGTM.
Thanks for the fix and test :)

**Review by Williambraecky — APPROVED** [review]
Thanks for the fix!

robodoo r+
